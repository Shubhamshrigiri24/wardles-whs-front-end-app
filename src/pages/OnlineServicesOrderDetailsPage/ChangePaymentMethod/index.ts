export { default, ChangePaymentMethod } from "./ChangePaymentMethod";
