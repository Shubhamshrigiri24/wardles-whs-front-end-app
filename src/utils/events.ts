export default {
  online: {
    consultation: {
      started: "Consultation started", // as the consultation is being loaded
      answeringQuestion: `Consultation answering question`, // when a user begins answering a question
      completed: "Consultation complete", // when the consultation is finished (when tripetto starts showing the epilogue)
    },
    productSelection: "Select product",
  },
  order: {
    tapOnSubscription: "tap on subscription",
    tapOnNoSubscription: "tap on no subscription",
    subscriptionMoreInfo: "tap on subscription more info",
    selectProductWithSubscription:
      "tap on continue to checkout with subscription",
    selectProductWithoutSubscription:
      "tap on continue to checkout one off purchase",
    deliveryDetails: "delivery details",
    paymentSuccess: "payment complete",
    confirmation: "confirmation complete",
  },
  subscription: {
    viewHealthStatusConfirmation: "view health status confirmation",
    submitHealthStatusConfirmation: "submit health status confirmation",
    cancel: "cancel subscription",
  },
  registration: {
    customer: "customer registration",
  },
};

export const eventsBuilder = {
  consultation: {
    start: "Started consultation",
    answer: "Question answered",
    ineligible: "Ineligible",
    complete: "Consultation Complete",
    referral: "Pharmacist Referral",
    referralSubmited: "Pharmacist Referral Submited",
    referralLink: "Pharmacist Referral Link",
  },
  productSelect: {
    productCardSelectButton: (product?: any) => `${product} selected`,
    packSelected: (
      productName?: string,
      productStrength?: string,
      productPackSize?: string,
      isSubscription?: boolean
    ) =>
      `${productName} ${productStrength}mg ${productPackSize} ${
        isSubscription ? "with" : "without"
      } subscription selected`,
    seeMoreAboutSubscription: "Clicked subscription info",
    choseProduct: "Chose product",
    choseProductWithSubscription: "Chose product with subscription",
  },
  basket: {
    discountApplied: "Discount Code Applied",
    discountFailed: "Discount Code Failed",
  },
  loginOptions: {
    signIn: "Chose Sign In",
    signInWithSub: "Chose Sign In With Subscription",
    signInComplete: "Sign in complete",
    customerLogin: "Services: customer login",
    register: "Chose Register",
    registerWithSub: "Chose Register With Subscription",
    guestCheckout: "Chose guest checkout",
    guestCheckoutWithSub: "Chose guest checkout with subscription",
    guestShowModal: "Guest checkout: Show subscription modal",
    guestRemoveSub: "Guest checkout: Removed subscription",
    guestBackToCheckout: "Guest checkout: Chose back to checkout option",
    viewTermsAndConditions: "Viewed terms and conditions",
  },
  orderDetailsForm: {
    newUser: {
      postcodeLookup: "Address: Postcode lookup",
      postcodeFailed: "Address: Postcode lookup failed",
      gpLookup: "GP lookup",
      gpLookupFailed: "GP lookup failed",
      accountDetailsSubmitted: "Account details: submitted",
      accountDetailsValidationError: "Account details: error",
    },
    existingUser: {
      postcodeLookup: "Delivery details: Postcode lookup",
      postcodeFailed: "Delivery details: Postcode lookup failed",
      deliveryDetailsSubmitted: "Delivery details: Submitted",
      deliveryDetailsValidationError: "Delivery details: Error",
      addressChangeModal: "Address change modal",
      addressChangeSubmit: "New delivery address submitted",
    },
  },
  account: {
    dashboard: "Account dashboard",
    cancelSubscriptionLink: "Chose cancel subscription",
    cancelSubscriptionConfirm: "Confirmed subscription cancellation",
    orderDetails: "Order details",
    subscriptionOrderDetails: "Subscription order details",
  },
  payment: {
    paymentComplete: "Payment complete",
    paymentWithSubcomplete: "Subscription payment complete",
    paymentFailed: "Payment failed",
    changePaymentMethod: "Chose change payment method",
    paymentMethodUpdate: "Updated payment method",
    paymentMethodUpdateFailed: "Payment update failed",
  },
  orderConfirmation: {
    showConfirmationScreen: "Order confirmed",
    showConfirmationScreenWithSub: "Subscription order confirmed",
  },
};
