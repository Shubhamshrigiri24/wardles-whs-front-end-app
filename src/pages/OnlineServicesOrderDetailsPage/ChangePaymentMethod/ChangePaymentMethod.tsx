import React, {
  FC,
  useState,
  useLayoutEffect,
  useRef,
  useCallback,
} from "react";
import { makeStyles, Theme } from "@material-ui/core";
import { Typography, Button, Alert, wellColors } from "@welldigital/components";
import { generateBraintreeStyles } from "pages/OnlineServicesCheckoutPaymentPage/styles";
import ChevronRight from "@material-ui/icons/ChevronRight";
import BraintreeWebDropIn from "braintree-web-drop-in";
import SubscriptionsApi from "utils/api/SubscriptionsAPI";
import { analytics } from "@welldigital/ui-common";
import { eventsBuilder } from "utils/events";

const useStyles = makeStyles((theme: Theme) => ({
  title: {
    fontSize: "34px",
    color: wellColors.elixir[900],
  },
  subTitle: {
    fontSize: "18px",
    fontWeight: theme.typography.fontWeightMedium,
    color: wellColors.elixir[300],
  },

  flexContainer: {
    display: "flex",
    flexWrap: "wrap",
    justifyContent: "center",
    [theme.breakpoints.up("sm")]: {
      flexWrap: "nowrap",
    },
  },
  braintreeContainer: {
    marginBottom: "16px",
    ...generateBraintreeStyles(theme),
  },
  braintreeInitSuccess: {
    minHeight: "270px",
  },
}));

export type ChangePaymentMethodProps = {
  brainTreeToken: string;
  subscriptionId: string;
  threeDSecure: any;
  onBack: () => void;
  performPageDataRefetch: () => void;
};

const braintreeContainerId = "braintree-change-payment-method";

export const ChangePaymentMethod: FC<ChangePaymentMethodProps> = ({
  brainTreeToken,
  threeDSecure,
  subscriptionId,
  onBack,
  performPageDataRefetch,
}) => {
  const classes = useStyles();
  const [changePaymentError, setChangePaymentError] = useState("");
  const [isChangingPayment, setIsChangingPayment] = useState(false);
  const [brainTreeInitError, setBraintreeInitError] = useState("");
  const [isButtonDisabled, setIsButtonDisabled] = useState(true);

  const braintreeInstance = useRef<BraintreeWebDropIn.Dropin | null>(null);

  const handlePaymentChangeTracking = useCallback((errorMessage?: string) => {
    analytics.trackEvent({
      flow: "subscription",
      event:
        eventsBuilder.payment[
          errorMessage ? "paymentMethodUpdateFailed" : "paymentMethodUpdate"
        ],
      metadata: {
        ...(errorMessage ? { error: errorMessage } : {}),
      },
    });
  }, []);

  const onChangePaymentMethod = useCallback(async () => {
    setIsChangingPayment(true);
    if (changePaymentError) {
      setChangePaymentError("");
    }
    try {
      const methodPayload =
        await braintreeInstance.current?.requestPaymentMethod({
          threeDSecure,
        });

      if (!methodPayload) {
        throw new Error("Error on updating payment method");
      }

      if (!methodPayload.liabilityShifted) {
        braintreeInstance.current?.clearSelectedPaymentMethod();
        setChangePaymentError(
          "We were unable to complete your 3DSecure check, please try a different payment method"
        );
        handlePaymentChangeTracking("3DSecure check failed");
        setIsChangingPayment(false);
        return;
      }

      await SubscriptionsApi.updateSubscriptionPayment(subscriptionId, {
        paymentNonce: methodPayload.nonce,
        paymentMethodDetails: {
          last4: (methodPayload.details as Record<string, any>).lastFour,
        },
      });

      handlePaymentChangeTracking();
      setIsChangingPayment(false);
      onBack();
      performPageDataRefetch();
    } catch (err) {
      console.error(err);
      braintreeInstance.current?.clearSelectedPaymentMethod();
      setChangePaymentError("Error on updating payment method");
      handlePaymentChangeTracking("Error on updating payment method");
      setIsChangingPayment(false);
    }
  }, [
    subscriptionId,
    threeDSecure,
    changePaymentError,
    onBack,
    performPageDataRefetch,
    handlePaymentChangeTracking,
  ]);

  useLayoutEffect(() => {
    window.scrollTo(0, 0);
    async function createInstance() {
      braintreeInstance.current = await BraintreeWebDropIn.create({
        container: `#${braintreeContainerId}`,
        authorization: brainTreeToken,
        threeDSecure: true,
        vaultManager: true,
        card: {
          overrides: {
            styles: {
              input: {
                padding: "16px",
                color: wellColors.elixir[400],
              },
              ".number": {
                "font-family": "monospace",
                "font-weight": 500,
                "font-size": "18px",
              },
              "input:focus": {
                color: wellColors.elixir[400],
              },
            },
          },
        },
      });

      if (braintreeInstance.current.isPaymentMethodRequestable()) {
        setIsButtonDisabled(false);
      }

      braintreeInstance.current?.on("paymentMethodRequestable", () => {
        // this should always be false because when we have token with saved methods the event.paymentMethodIsSelected is true and when we don't have a payment method saved then  event.paymentMethodIsSelected is false , but in both cases we need to be able to press the button and is not reliable -- see braintree docs for further clarifications if needed
        //https://developer.paypal.com/braintree/docs/guides/drop-in/customization
        setIsButtonDisabled(false);
      });
      braintreeInstance.current?.on("noPaymentMethodRequestable", () => {
        setIsButtonDisabled(true);
      });
    }

    createInstance().catch((err) => {
      console.error("Could not begin change payment!", err);
      setBraintreeInitError(err.message ?? "Braintree init error!");
    });

    return () => {
      braintreeInstance.current?.teardown();
    };
  }, [brainTreeToken]);

  return (
    <div>
      <Typography className={classes.title} variant={"h2"} spacingAfter={2}>
        Payment
      </Typography>
      {brainTreeInitError && (
        <Alert message={brainTreeInitError} type={"error"} />
      )}
      {!brainTreeInitError && (
        <>
          <Typography className={classes.subTitle} spacingAfter={2}>
            Your card details will be saved securely and we will apply to take
            payment for your subscription every month.
          </Typography>
          <div className={classes.braintreeContainer}>
            <div id={braintreeContainerId} />
          </div>
          {changePaymentError && (
            <Alert
              type={"error"}
              spacingAfter={2}
              message={changePaymentError}
            />
          )}
          <div className={classes.flexContainer}>
            <Button variant={"text"} color={"default"} onClick={onBack}>
              <Typography style={{ color: wellColors.elixir[900] }}>
                Back
              </Typography>
            </Button>
            <Button
              fullWidth
              variant={"contained"}
              size={"large"}
              data-testid={"changePayment/payment-button"}
              color={"primary"}
              loading={isChangingPayment}
              disabled={isButtonDisabled}
              onClick={onChangePaymentMethod}
              endIcon={<ChevronRight />}
            >
              Change payment method
            </Button>
          </div>
        </>
      )}
    </div>
  );
};

export default ChangePaymentMethod;
