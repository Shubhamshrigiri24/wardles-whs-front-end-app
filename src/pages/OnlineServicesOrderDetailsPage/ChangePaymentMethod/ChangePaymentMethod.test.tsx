import React from "react";
import {
  render,
  cleanup,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import ChangePaymentMethod, {
  ChangePaymentMethodProps,
} from "./ChangePaymentMethod";
import mockBraintree from "braintree-web-drop-in";
import mockSubscriptionsApi from "utils/api/SubscriptionsAPI";
import { EventEmitter } from "events";

const mockOnBack = jest.fn();
const mockPerformDataRefetch = jest.fn();
const braintreeInstance = Object.assign(new EventEmitter(), {
  requestPaymentMethod: jest.fn(),
  isPaymentMethodRequestable: jest.fn(),
  clearSelectedPaymentMethod: jest.fn(),
  teardown: jest.fn(),
});

const noop = () => {};
Object.defineProperty(window, "scrollTo", { value: noop, writable: true });

jest.mock("../../../utils/api/SubscriptionsAPI", () => ({
  updateSubscriptionPayment: jest.fn(),
}));

jest.mock("braintree-web-drop-in", () => ({
  create: jest.fn(),
}));

beforeEach(() => {
  jest.resetAllMocks();
  (mockBraintree.create as jest.Mock).mockReturnValue(braintreeInstance);
});

const makeProps = (
  props?: Partial<ChangePaymentMethodProps>
): ChangePaymentMethodProps => {
  return {
    brainTreeToken: "12345678910",
    subscriptionId: "abcdefg",
    threeDSecure: {},
    onBack: mockOnBack,
    performPageDataRefetch: mockPerformDataRefetch,
    ...props,
  };
};

describe("<ChangePaymentMethod />", () => {
  it("should match snapshot", () => {
    const props = makeProps({});
    const { container } = render(<ChangePaymentMethod {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should create drop in instance", async () => {
    const props = makeProps();
    const braintreeEventsSpy = jest.spyOn(braintreeInstance, "on");
    await render(<ChangePaymentMethod {...props} />);
    expect(braintreeEventsSpy).toBeCalledTimes(2);
    expect(braintreeEventsSpy.mock.calls[0][0]).toEqual(
      "paymentMethodRequestable"
    );
    expect(braintreeEventsSpy.mock.calls[1][0]).toEqual(
      "noPaymentMethodRequestable"
    );
    expect(braintreeInstance.isPaymentMethodRequestable).toBeCalledTimes(1);
  });

  it("should destroy braintree instance on unmount", async () => {
    const props = makeProps();
    (mockBraintree.create as jest.Mock).mockReturnValue(braintreeInstance);
    await render(<ChangePaymentMethod {...props} />);
    await cleanup();
    expect(braintreeInstance.teardown).toBeCalledTimes(1);
  });

  it("should change payment method successfully", async () => {
    const props = makeProps();
    const braintreeListenerSpy = jest.spyOn(braintreeInstance, "on");
    braintreeInstance.requestPaymentMethod.mockReturnValue({
      liabilityShifted: true,
      details: {
        lastFour: "1111",
      },
    });
    const { getByTestId } = await render(<ChangePaymentMethod {...props} />);
    const payButton = getByTestId("changePayment/payment-button");
    act(() => {
      const braintreePaymentRequestableCallback =
        braintreeListenerSpy.mock.calls[0][1];
      braintreePaymentRequestableCallback();
    });
    await act(async () => {
      fireEvent.click(payButton);
    });

    expect(braintreeInstance.requestPaymentMethod).toBeCalledTimes(1);
    expect(mockSubscriptionsApi.updateSubscriptionPayment).toBeCalledTimes(1);
    expect(mockOnBack).toBeCalledTimes(1);
    expect(mockPerformDataRefetch).toBeCalledTimes(1);
  });

  it("should show error if requesting a payment method fails", async () => {
    const props = makeProps();
    const braintreeListenerSpy = jest.spyOn(braintreeInstance, "on");
    braintreeInstance.requestPaymentMethod.mockReturnValue(null);

    const { getByText, getByTestId } = await render(
      <ChangePaymentMethod {...props} />
    );

    act(() => {
      const braintreePaymentRequestableCallback =
        braintreeListenerSpy.mock.calls[0][1];
      braintreePaymentRequestableCallback();
    });
    const payButton = getByTestId("changePayment/payment-button");
    await act(async () => {
      fireEvent.click(payButton);
    });

    expect(getByText(/Error on updating payment method/i)).toBeInTheDocument();
    expect(braintreeInstance.clearSelectedPaymentMethod).toBeCalledTimes(1);
  });

  it("should show error when 3dsecure auth fails", async () => {
    const props = makeProps();
    const braintreeListenerSpy = jest.spyOn(braintreeInstance, "on");
    braintreeInstance.requestPaymentMethod.mockReturnValue({
      liabilityShifted: false,
    });
    const { getByText, getByTestId } = await render(
      <ChangePaymentMethod {...props} />
    );
    const payButton = getByTestId("changePayment/payment-button");
    act(() => {
      const braintreePaymentRequestableCallback =
        braintreeListenerSpy.mock.calls[0][1];
      braintreePaymentRequestableCallback();
    });
    await act(async () => {
      fireEvent.click(payButton);
    });

    expect(
      getByText(
        /We were unable to complete your 3DSecure check, please try a different payment method/i
      )
    ).toBeInTheDocument();
    expect(braintreeInstance.clearSelectedPaymentMethod).toBeCalledTimes(1);
  });
});
