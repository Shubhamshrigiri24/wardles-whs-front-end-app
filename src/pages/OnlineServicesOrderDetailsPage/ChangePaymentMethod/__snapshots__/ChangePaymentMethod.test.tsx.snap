// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePaymentMethod /> should match snapshot 1`] = `
<div>
  <div>
    <h2
      class="MuiTypography-root makeStyles-title-1 MuiTypography-h2"
    >
      Payment
    </h2>
    <div
      class="makeStyles-spacing2-10 makeStyles-spacing2Responsive-11"
      data-testid="spacing/root"
    />
    
    <p
      class="MuiTypography-root makeStyles-subTitle-2 MuiTypography-body1"
    >
      Your card details will be saved securely and we will apply to take payment for your subscription every month.
    </p>
    <div
      class="makeStyles-spacing2-10"
      data-testid="spacing/root"
    />
    <div
      class="makeStyles-braintreeContainer-4"
    >
      <div
        id="braintree-change-payment-method"
      />
    </div>
    
    <div
      class="makeStyles-flexContainer-3"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <div
            class="makeStyles-children-34"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
              style="color: rgb(12, 22, 31);"
            >
              Back
            </p>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled MuiButton-fullWidth Mui-disabled"
        data-testid="changePayment/payment-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <div
            class="makeStyles-children-34"
          >
            Change payment method
          </div>
          <span
            class="MuiButton-endIcon makeStyles-endIcon-36 MuiButton-iconSizeLarge"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
